tracing-tracy/src/config.rs

```diff
@@ -0,0 +1,145 @@
+use client::Client;
+use tracing_subscriber::fmt::format::DefaultFields;
+use tracing_subscriber::fmt::FormatFields;
+
+/// Configuration of the [`TracyLayer`] behaviour.
+///
+/// For most users [`DynamicConfig`] is going to be a good default option, however advanced users
+/// can implement this trait manually to achieve better performance through constant evaluation,
+/// to override the formatter used or to otherwise modify the behaviour of `TracyLayer` in ways
+/// that are not exposed via the `DynamicConfig` type.
+///
+/// # Examples
+///
+/// ## Implementation with compile-time configuration
+///
+/// ```
+/// #[derive(Default)]
+/// struct ConstantTracyConfig {
+///     formatter: tracing_subscriber::fmt::format::DefaultFields,
+/// }
+///
+/// impl tracing_tracy::Config for ConstantTracyConfig {
+///     type Formatter = tracing_subscriber::fmt::format::DefaultFields;
+///     fn formatter(&self) -> &Self::Formatter { &self.formatter }
+///     fn stack_depth(&self) -> u16 { 0 } // Same as the default trait impl.
+///     fn format_fields_in_zone_name(&self) -> bool { true } // Same as the default trait impl.
+/// }
+/// ```
+///
+/// With this sort of setup the compiler will be able to inline calls to `stack_depth` and
+/// `format_fields_in_zone_name` and optimize accordingly.
+pub trait Config {
+    type Formatter: for<'writer> FormatFields<'writer> + 'static;
+
+    /// Use a custom field formatting implementation.
+    fn formatter(&self) -> &Self::Formatter;
```

Comment:
I wish there was a way to specify a default here. Sadly not much progress: https://rust-lang.github.io/rfcs/2532-associated-type-defaults.html

Reply:
I wish it was possible too, however even with ATDs it wouldn’t be possible as the method returns a reference to `&self` and traits do not have access to fields. Achieving this would need something like https://github.com/rust-lang/rfcs/pull/1546.